Flash Recovery Boot

mlvisaya opened the pull request:
- Load Caliptra firmware, manifest, MCU RT from the active partition of the flash and push to the recovery interface instead of the BMC pushing these images.

- Use the MCU I3C Recovery interface instead of the Caliptra Recovery Interface.

- Implement the BMC Recovery State Machine in MCU ROM

- Dependent on https://github.com/chipsalliance/caliptra-sw/pull/2269/

helloxiling commented on emulator/caliptra/src/caliptra.rs:
```diff
@@ -197,6 +199,12 @@ pub fn start_caliptra(
         })
     };
 
+    if args_use_mcu_recovery_interface {
```
nit: we can remove those debug prints.

helloxiling commented on rom/src/lib.rs:
```diff
@@ -19,11 +19,39 @@ pub use fuses::*;
+    Unknown,
+}
+
+pub trait FlashPartitionIntf {
```
can we reuse `platforms/emulator/rom/src/flash/flash_api.rs`, which is defined as API for MCU ROM to access flash partition?

helloxiling replied:
We may restructure the flash driver and API to enable better sharing between common_rom and emulator_rom.

mlvisaya replied:
I will restructure a bit, so that there will be  FlashStorage Interface in common ROM, and the flash_api will use that generic interface. The implementation (EmulatedFlashCtrl) will implement the FlashStorage interface in ROM Emulator, and we can use a default implementation of the flash_api

helloxiling commented on platforms/emulator/rom/src/riscv.rs:
```diff
@@ -91,3 +175,199 @@ fn exit_rom() -> ! {
     }
 }
+
+pub struct RomEmulatedFlash<'a> {
```
Please check if we can reuse `platforms/emulator/rom/src/flash/flash_api.rs`.

mlvisaya replied:
removed this, and just used flash_api.

helloxiling commented on platforms/emulator/rom/src/riscv.rs:
```diff
@@ -91,3 +175,199 @@ fn exit_rom() -> ! {
+    }
+}
+
+pub struct FlashBootCfg<'a> {
```
Would it make sense to relocate the `FlashBootCfg trait implementation under the `flash` module for better organization?

mlvisaya replied:
yes, will move this under flash

helloxiling commented on platforms/emulator/runtime/userspace/api/caliptra-api/src/flash_boot_cfg.rs:
This file seems closely tied to the image loading phase of the boot flow. Would it make sense to move this file into the `image_loading` module?

mlvisaya replied:
yes, it can be placed under an image_loading module, however, since this is a sample implementation of BootConfig trait, it needs to be under emulator. The image_loading module is currently in: 

runtime/userspace/api/caliptra-api/src/image_loading

I can create the same directory under platforms/emulator, and put it here.

helloxiling commented on registers/generated-emulator/src/i3c.rs:
```diff
@@ -2,9 +2,19 @@
 #[allow(unused_imports)]
 use tock_registers::interfaces::{Readable, Writeable};
 pub trait I3cPeripheral {
+    fn register_event_channels(
```
We should include this routine in `fn emu_make_peripheral_trait` line 449 `tokens.extend` within `xtask/src/registers.rs`, then run `cargo xtask registers-autogen`. That way, it gets applied to all peripherals automatically. I ran into this earlier when I tried adding trait functions manually.

helloxiling commented on platforms/emulator/rom/src/riscv.rs:
```diff
@@ -44,7 +59,76 @@ pub extern "C" fn rom_entry() -> ! {
+    #[cfg(feature = "test-flash-based-boot")]
+    {
+        // Initialize the flash controller for testing purposes
+        let mut partition_table_flash =
```
nit: we can just name it as `primary_flash_ctrl`.

helloxiling commented on platforms/emulator/rom/src/riscv.rs:
```diff
@@ -44,7 +59,76 @@ pub extern "C" fn rom_entry() -> ! {
+        let partition_a_flash = EmulatedFlashCtrl::initialize_flash_ctrl(PRIMARY_FLASH_CTRL_BASE);
+        let partition_b_flash = EmulatedFlashCtrl::initialize_flash_ctrl(SECONDARY_FLASH_CTRL_BASE);
+
+        let mut partition_a_driver = FlashPartition::new(
```
Nit: just use `partition_a` to keep it simple.

mlvisaya commented on rom/src/i3c.rs:
```diff
@@ -166,4 +166,35 @@ impl I3c {
                 HcControl::BusEnable::CLEAR,
         );
     }
+    pub fn read_indirect_fifo_status_2(&self) -> u32 {
```
removed this, and use the given i3c peripheral instead

mlvisaya commented:
> Thanks for addressing the comments. The updated patch looks pretty good! I do have a quick suggestion about the flash module layout in ROM:
> 
> ```
> rom/
>   src/
>     flash/
>       mod.rs              
>       hil.rs               // Defines FlashStorage trait for logical flash driver
>       flash_partition.rs   // Implements FlashPartition that serves as flash API
> 
> platforms/
>   emulator/
>     rom/
>       flash/
>         mod.rs
>         flash_drv.rs       // Implements page-based flash_ctrl driver and FlashStorage trait over it
> ```
> 
> The reason is that we want to maintain consistent flash file naming throughout the codebase and use a similar hierarchy structure as runtime flash stack: `flash_ctrl.rs` is used to represent page-based flash controller driver, and `flash_storage` is used to represent logical flash driver. Let's try to align that in ROM flash stack. Please let me know if this makes sense.

that makes sense. will make the naming consistent

helloxiling reviewed (approved):
Looks good to me.

swenson commented on Cargo.toml:
```diff
@@ -187,28 +189,28 @@ libtock_small_panic = { path = "runtime/userspace/libtock/panic_handlers/small_p
-caliptra-test-harness-types = { git = "https://github.com/chipsalliance/caliptra-sw", rev = "7fb2e579cb2a932fabf804c50e83c9ba78574836" }
-ureg = { git = "https://github.com/chipsalliance/caliptra-sw", rev = "7fb2e579cb2a932fabf804c50e83c9ba78574836" }
-dpe = { git = "https://github.com/chipsalliance/caliptra-sw", rev = "7fb2e579cb2a932fabf804c50e83c9ba78574836", default-features = false, features = ["dpe_profile_p384_sha384"] }
+caliptra-api = { git = "https://github.com/chipsalliance/caliptra-sw", rev = "b0f0705a70fcadc3192f8c0fdb2c80902bff2632" }
```
We'll need to update this after https://github.com/chipsalliance/caliptra-sw/pull/2269 is merged

mlvisaya replied:
yes, will update caliptra-sw dependency with latest SHA

swenson commented on emulator/periph/src/i3c.rs:
```diff
@@ -188,9 +229,269 @@ impl I3c {
+            return Err(caliptra_emu_bus::BusError::StoreAddrMisaligned);
+        }
+        match addr {
+            0x00c..0x010 => {
```
To simplify these, you can eliminate the second half, e.g., `0xc => {`.

It would be invalid for the write to be to `0xd`, `0xe`, or `0xf` in any case.

mlvisaya replied:
yeah, makes sense. i'll remove it.

swenson commented on emulator/periph/src/i3c.rs:
```diff
@@ -291,10 +592,48 @@ impl I3cPeripheral for I3c {
+            .i3c_ec_soc_mgmt_if_rec_intf_cfg
+            .reg
+            .read(RecIntfCfg::RecIntfBypass);
+        if bypass_cfg == I3C_REC_INT_BYPASS_I3C_CORE {
```
A thought: this is only supported by 2.1 hardware. We need to decide how we want to handle 2.0 vs. 2.1 hardware in this repository.

I think we'll want either a command-line flag in the emulator to select the hardware version or a feature flag.

We have an SoC firmware register that can be used to determine this as well: https://chipsalliance.github.io/caliptra-rtl/main/internal-regs/?p=clp.soc_ifc_reg.CPTRA_HW_REV_ID

mlvisaya replied:
i could pass a command-line flag with the HW version.

swenson commented:
Sorry for the merge conflicts!

mlvisaya commented on emulator/periph/src/i3c.rs:
```diff
@@ -57,6 +57,7 @@ pub struct I3c {
     _error_irq: Irq,
     /// Notification interrupt
     notif_irq: Irq,
+    hw_revision: String,
```
that's a good idea. I'll use semver
